asmcomp/comballoc.ml

```diff
@@ -34,25 +34,23 @@ let rec combine i allocstate =
       (i, allocated_size allocstate)
   | Iop(Ialloc { words = sz; _ }) ->
       begin match allocstate with
-        No_alloc ->
-          let (newnext, newsz) =
-            combine i.next (Pending_alloc(i.res.(0), sz)) in
-          (instr_cons_debug (Iop(Ialloc {words = newsz; spacetime_index = 0;
-              label_after_call_gc = None; }))
-            i.arg i.res i.dbg newnext, 0)
-      | Pending_alloc(reg, ofs) ->
-          if ofs + sz < Config.max_young_wosize * Arch.size_addr then begin
-            let (newnext, newsz) =
-              combine i.next (Pending_alloc(reg, ofs + sz)) in
-            (instr_cons (Iop(Iintop_imm(Iadd, ofs))) [| reg |] i.res newnext,
+      | Pending_alloc(reg, totalsz)
+          when totalsz + sz < Config.max_young_wosize * Arch.size_addr ->
+         let (newnext, newsz) =
+              combine i.next (Pending_alloc(i.res.(0), totalsz + sz)) in
+            (instr_cons_debug (Iop(Iintop_imm(Iadd, -sz)))
+               [| reg |] i.res i.dbg newnext,
              newsz)
-          end else begin
-            let (newnext, newsz) =
-              combine i.next (Pending_alloc(i.res.(0), sz)) in
-            (instr_cons_debug (Iop(Ialloc { words = newsz; spacetime_index = 0;
-                label_after_call_gc = None; }))
-              i.arg i.res i.dbg newnext, ofs)
-          end
+      | _ ->
+         let (newnext, newsz) = combine i.next (Pending_alloc(i.res.(0), sz)) in
+         let newnext =
+           if newsz = sz then newnext
```

Comment:
I think this might be more immediately clear with something like this, having done the rename of `newsz`:
```
let required_offset = totalsz - sz in
if required_offset <= 0 then newnext
else instr_cons_debug ... required_offset ...
```